aten/src/ATen/native/cuda/Shape.cu

```diff
@@ -379,12 +372,27 @@ Tensor& cat_out_cuda(Tensor& out, TensorList inputs, int64_t dimension) {
     [=](const Tensor& t) {
       return !t.defined() || t.is_contiguous(memory_format);
     });
+  ScalarType firstType = inputs[0].scalar_type();
```

Comment:
This section seems a bit complicated, can this be written as something like `canCast( result_type( inputs ), out )` ?

Reply:
The comparison of all types being the same is for activating the fast path. I think canCast alone does not provide enough information for this so we need to explicitly do the check 😔.
However, I rewrote the loop right after with your suggestion 😃 which makes that part a bit more readable

Reply:
> I think canCast alone does not provide enough information

Why not? Effectively this is what TensorIterator does, just calculating the common dtype and making sure that's castable to the output.

Reply:
The problem I saw here is that in some conditions, an optimized path is executed `parallel_cat` https://github.com/pytorch/pytorch/blob/5c925e7ca53c1792f2d71525a2cf5fc4daacf371/aten/src/ATen/native/cuda/Shape.cu#L398 
the way this `parallel_cat` works needs all the tensors to be the same input type as it 
relies on some templating and macros that seems tricky to change.

Checking this condition only with `canCast` is not safe because it does not provide this information that is needed for the optimized version to work, as canCast tell us if the type promotion is doable but not if all the types are the same.

Reply:
I decoupled the two checks, I now do the canCast at the beginning of the function, and here I just check the types to enable the fast implementation.